frontend/model/contracts/mailbox.js

```diff
@@ -1,52 +1,162 @@
 'use strict'
 
 import sbp from '@sbp/sbp'
-import { Vue } from '@common/common.js'
-import { mailType } from './shared/types.js'
-import { objectOf, string, object, optional } from '~/frontend/model/contracts/misc/flowTyper.js'
+import { Vue, L } from '@common/common.js'
+import { merge } from './shared/giLodash.js'
+import { leaveChatRoom } from './shared/functions.js'
+import { objectOf, string, boolean, optional } from '~/frontend/model/contracts/misc/flowTyper.js'
+import { logExceptNavigationDuplicated } from '~/frontend/views/utils/misc.js'
 
 sbp('chelonia/defineContract', {
   name: 'gi.contracts/mailbox',
   metadata: {
-    // TODO: why is this missing the from username..?
     validate: objectOf({
-      createdDate: string
+      createdDate: string, // action created date
+      username: optional(string), // action creator
+      identityContractID: optional(string) // action creator identityContractID
     }),
     create () {
+      if (!sbp('state/vuex/state').loggedIn) {
+        return { createdDate: new Date().toISOString() }
+      }
+      const { username, identityContractID } = sbp('state/vuex/state').loggedIn
       return {
-        createdDate: new Date().toISOString()
+        createdDate: new Date().toISOString(),
+        username,
+        identityContractID
       }
     }
   },
   actions: {
     'gi.contracts/mailbox': {
-      validate: object, // TODO: define this
+      validate: objectOf({
+        username: string
+      }),
       process ({ data }, { state }) {
-        for (const key in data) {
-          Vue.set(state, key, data[key])
+        const initialState = merge({
+          attributes: {
+            creator: data.username,
+            autoJoinAllowance: true
+          },
+          users: {}
+        }, data)
+        for (const key in initialState) {
+          Vue.set(state, key, initialState[key])
         }
-        Vue.set(state, 'messages', [])
       }
     },
-    'gi.contracts/mailbox/postMessage': {
-      validate: objectOf({
-        messageType: mailType,
-        from: string,
-        subject: optional(string),
-        message: optional(string),
-        headers: optional(object)
-      }),
-      process (message, { state }) {
-        state.messages.push(message)
+    'gi.contracts/mailbox/setAutoJoinAllowance': {
```

Comment:
What is this?

Also - it is typically better / simpler to have a single action for setting all attributes, instead of having an action for each attribute. For example, see `'gi.contracts/identity/setAttributes'`

Reply:
> What is this?

It's not in the requirement and there is no design for this yet. But I thought it could be better. No more syncing chatroom contracts if the user don't want new DM more.

> Also - it is typically better / simpler to have a single action for setting all attributes, instead of having an action for each attribute. For example, see 'gi.contracts/identity/setAttributes'

Yeah. That's great. Just like what we did in identity contract.